crates/matrix-sdk/src/send_queue.rs

```diff
@@ -353,30 +366,68 @@ impl RoomSendQueue {
                 }
 
                 Err(err) => {
-                    warn!(txn_id = %queued_event.transaction_id, "error when sending event: {err}");
-
-                    // In this case, we intentionally keep the event in the queue, but mark it as
-                    // not being sent anymore.
-                    queue.mark_as_not_being_sent(&queued_event.transaction_id).await;
-
-                    // Let observers know about a failure *after* we've marked the item as not
-                    // being sent anymore. Otherwise, there's a possible race where a caller might
-                    // try to remove an item, while it's still marked as being sent, resulting in a
-                    // cancellation failure.
-
-                    // Disable the queue for this room after an error.
-                    locally_enabled.store(false, Ordering::SeqCst);
+                    let (err, is_recoverable) = if let crate::Error::Http(http_err) = err {
+                        // A small trick to confuse borrowck (and humans too).
```

Comment:
Hmm, this smells. Aren't you in both cases where `characterize_retry_kind()` is used (which is clearly a contstructor for `RetryKind`, not sure why it's not under an `impl` block for `RetryKind`) immediately deconstructing or taking the error out of `RetryKind`?

Why does `RetryKind` then take ownership of the error in the first place? Furthermore, if we don't take ownership wouldn't something like this work much better:

```rust
pub(crate) enum RetryKind {
    Transient {
        err: HttpError,
        // This is used only for attempts to retry, so on non-wasm32 code (in the `native` module).
        #[cfg_attr(target_arch = "wasm32", allow(dead_code))]
        retry_after: Option<Duration>,
    },
    Permanent(HttpError),
}

impl HttpError {
    pub(crate) fn retry_kind(&self) -> RetryKind {
        ... body of `characterize_retry_kind()`
    }
}
```

Granted, `RetryKind` starts to look suspiciously similar to `RetryError` but perhaps we just don't want to include that type in our public API.

Reply:
Edited my comment, I forgot to remove the error field from `RetryKind`. Should now be possible.

Reply:
Great idea, thanks, done.